### ISSUE TITLE
No diff is detected and applied if we enable server side apply(ssa) and change a listMapKey of listType struct list

### ISSUE THREAD

**Issue by loveRhythm1990** [pulumi/pulumi-kubernetes#3189]
### What happened?

`ClusterSpec` is a crd spec, it has a `CNSets` field definition like that.
```
type ClusterSpec struct {
	// +listType=map
	// +listMapKey=name
	CNSets []Component `json:"cnSets,omitempty"`
}

type Component struct {
      Name string `json:"name"`
      // some other complex fields
}
```
part of `Component` fields are managed by pulumi (include `name` field) by `apply`, other fields are managed by our controller by `update`.

when we update Component.Name in pulumi, it is listMapKey, `pulumi preview` shows no diff, and not applied to K8s.

the result of  `pulumi stack export`  likes that, the important part is that the `outputs` is not same with `inputs` on name field
```
inputs:
    name: default-localpv
outputs:
    name: default
```

detail is following, please just focus `cnset.name` field.
```
            {
                "urn": "urn:pulumi:new-dev::cos-resources::kubernetes:yaml:ConfigFile$kubernetes:core.<org>/v1alpha1:Cluster::issue2612-test",
                "custom": true,
                "id": "issue2612-test",
                "type": "kubernetes:core.<org>/v1alpha1:Cluster",
                "inputs": {
                    "apiVersion": "core.<org>/v1alpha1",
                    "kind": "Cluster",
                    "metadata": {
                        "name": "issue2612-test"
                    },
                    "spec": {
                        "accountID": "issue2612",
                        "cnSets": [
                            {
                                "managed": {
                                    "componentBackend": "Pod",
                                },
                                "name": "default-localpv",
                                "profile": "cn.localpv",
                                "replicas": 0
                            }
                        ],
                        "endpoint": {
                            "mode": "Shared",
                            "whitelistIPRanges": [
                                "0.0.0.0/0"
                            ]
                        },
                        "mainClusterRef": "freetier-01",
                        "managementPolicy": "Manual",
                        "state": "Active",
                        "unitName": "unit-cn-hangzhou",
                        "version": "dev-54fc6a89b"
                    }
                },
                "outputs": {
                    "__fieldManager": "pulumi-kubernetes-41eb1472",
                    "__initialApiVersion": "core.<org>/v1alpha1",
                    "__inputs": {
                        "apiVersion": "core.m<org>/v1alpha1",
                        "kind": "Cluster",
                        "metadata": {
                            "labels": {
                                "<org>/role": "internal-test"
                            },
                            "name": "issue2612-test"
                        },
                        "spec": {
                            "accountID": "issue2612",
                            "cnSets": [
                                {
                                    "managed": {
                                        "componentBackend": "Pod",
                                    },
                                    "name": "default",
                                    "profile": "cn.localpv",
                                    "replicas": 0
                                }
                            ],
                            "endpoint": {
                                "mode": "Shared",
                                "whitelistIPRanges": [
                                    "0.0.0.0/0"
                                ]
                            },
                            "mainClusterRef": "freetier-01",
                            "managementPolicy": "Manual",
                            "state": "Active",
                            "unitName": "unit-cn-hangzhou",
                            "version": "dev-54fc6a89b"
                        }
                    },
                    "apiVersion": "core.<org>/v1alpha1",
                    "kind": "Cluster",
                    "metadata": {
                        "creationTimestamp": "2024-08-27T11:56:20Z",
                        "generation": 3,
                        "labels": {
                            "<org>/role": "internal-test"
                        },
                        "managedFields": [
                            {
                                "apiVersion": "core.<org>/v1alpha1",
                                "fieldsType": "FieldsV1",
                                "fieldsV1": {
                                    "f:metadata": {
                                        "f:labels": {
                                            "f:<org>/role": {}
                                        }
                                    },
                                    "f:spec": {
                                        "f:accountID": {},
                                        "f:cnSets": {
                                            "k:{\"name\":\"default\"}": {
                                                ".": {},
                                                "f:managed": {
                                                    "f:componentBackend": {},
                                                    "f:config": {}
                                                },
                                                "f:name": {},
                                                "f:profile": {},
                                                "f:replicas": {}
                                            }
                                        },
                                        "f:endpoint": {
                                            "f:mode": {},
                                            "f:whitelistIPRanges": {}
                                        },
                                        "f:mainClusterRef": {},
                                        "f:managed": {
                                            "f:imageRepository": {}
                                        },
                                        "f:managementPolicy": {},
                                        "f:state": {},
                                        "f:unitName": {},
                                        "f:unitSelector": {},
                                        "f:version": {}
                                    }
                                },
                                "manager": "pulumi-kubernetes-41eb1472",
                                "operation": "Apply",
                                "time": "2024-08-27T11:56:20Z"
                            },
                            {
                                "apiVersion": "core.<org>/v1alpha1",
                                "fieldsType": "FieldsV1",
                                "fieldsV1": {
                                    "f:metadata": {
                                        "f:finalizers": {
                                            ".": {},
                                        }
                                    },
                                    "f:spec": {
                                        "f:cnSets": {
                                            "k:{\"name\":\"default\"}": {
                                                "f:managed": {
                                                    "f:nodeSelector": {
                                                        ".": {},
                                                        "f:<org>/cluster-pool": {},
                                                        "f:<org>/profile": {}
                                                    },
                                                    "f:resources": {
                                                        "f:limits": {
                                                            ".": {},
                                                            "f:cpu": {},
                                                            "f:memory": {}
                                                        },
                                                        "f:requests": {
                                                            ".": {},
                                                            "f:cpu": {},
                                                            "f:memory": {}
                                                        }
                                                    },
                                                    "f:storageClassName": {},
                                                    "f:storageSize": {},
                                                    "f:toleratedTaints": {}
                                                }
                                            }
                                        }
                                    }
                                },
                                "manager": "Go-http-client",
                                "operation": "Update",
                                "time": "2024-08-27T11:56:20Z"
                            },
                            {
                                "apiVersion": "core.<org>/v1alpha1",
                                "fieldsType": "FieldsV1",
                                "fieldsV1": {
                                    "f:status": {
                                        ".": {},
                                        "f:conditions": {},
                                        "f:endpoint": {},
                                        "f:phase": {}
                                    }
                                },
                                "manager": "Go-http-client",
                                "operation": "Update",
                                "subresource": "status",
                                "time": "2024-08-27T11:56:20Z"
                            }
                        ],
                        "name": "issue2612-test",
                        "resourceVersion": "147702716",
                        "uid": "2d4b9761-b52e-4259-8d6e-06fff48a827f"
                    },
                    "spec": {
                        "accountID": "issue2612",
                        "cnSets": [
                            {
                                "managed": {
                                    "componentBackend": "Pod",
                                    "nodeSelector": {
                                    },
                                    "resources": {
                                        "limits": {
                                            "cpu": "7",
                                            "memory": "28Gi"
                                        },
                                        "requests": {
                                            "cpu": "7",
                                            "memory": "28Gi"
                                        }
                                    },
                                    "storageClassName": "local-pv",
                                    "storageSize": "430Gi",
                                },
                                "name": "default",
                                "profile": "cn.localpv",
                                "replicas": 0
                            }
                        ],
                        "endpoint": {
                            "mode": "Shared",
                            "whitelistIPRanges": [
                                "0.0.0.0/0"
                            ]
                        },
                        "mainClusterRef": "freetier-01",
                        "managementPolicy": "Manual",
                        "state": "Active",
                        "unitName": "unit-cn-hangzhou",
                        "version": "dev-54fc6a89b"
                    },
                    "state": {},
                    "status": {
                        "conditions": [
                            {
                                "lastTransitionTime": "2024-08-27T11:56:20Z",
                                "message": "the object is reconciling",
                                "observedGeneration": 1,
                                "reason": "empty",
                                "status": "False",
                                "type": "Synced"
                            }
                        ],
                        "endpoint": {},
                        "phase": "Creating"
                    }
                },
                "parent": "urn:pulumi:new-dev::cos-resources::kubernetes:yaml:ConfigFile::issue2612-cluster",
                "provider": "urn:pulumi:new-dev::cos-resources::pulumi:providers:kubernetes::kubernetes-provider::e0c59a12-875b-4692-be43-d0938e88a7e0",
                "propertyDependencies": {
                    "apiVersion": [],
                    "kind": [],
                    "metadata": [],
                    "spec": []
                },
                "created": "2024-08-27T11:56:20.536754827Z",
                "modified": "2024-08-27T11:56:20.536754827Z"
            },

```     

### Example

like above

### Output of `pulumi about`

lr90@192 cos-resources % pulumi about
CLI          
Version      3.129.0
Go Version   go1.22.6
Go Compiler  gc

Plugins
KIND      NAME        VERSION
resource  alicloud    3.43.0
language  go          unknown
resource  kubernetes  3.30.2

Host     
OS       darwin
Version  14.5
Arch     x86_64

This project is written in go: executable='/Users/lr90/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.4.darwin-amd64/bin/go' version='go version go1.22.4 darwin/amd64'

Found no pending operations associated with new-dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/<org>
User           <org>
Organizations <org>
Token type     personal

Dependencies:
NAME                                            VERSION
github.com/pulumi/pulumi-alicloud/sdk/v3        v3.43.0
github.com/pulumi/pulumi-kubernetes/sdk/v3      v3.30.2
github.com/pulumi/pulumi/sdk/v3                 v3.78.1
k8s.io/api                                      v0.27.2
k8s.io/apimachinery                             v0.27.2

### Additional context

if we update the annotation of the CR, the name update will be apply to K8s.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by loveRhythm1990** [comment]
this issue bothers us a long time, Could you please help take a look, @lblackstone, @blamp, greatly appreciated

**Comment by rquitales** [comment]
Thank you for the detailed report on this issue, @loveRhythm1990. I noticed that you're currently using an older version of our Kubernetes provider. Upgrading to version 4 should resolve this issue. The problem you're experiencing stems from the behavior in version 3, where the provider always performs a diff between Pulumi program inputs rather than outputs, which is not the intended behavior you described. This was addressed in issue #2445 by updating the diffing logic to compare the last known cluster output against new inputs, which should resolve the problem. This fix is included in version 4 of our provider.

Please try upgrading to version 4 of the Pulumi Kubernetes provider and let us know if the issue persists. Thank you!

**Comment by loveRhythm1990** [comment]
> Thank you for the detailed report on this issue, [@loveRhythm1990](https://github.com/loveRhythm1990). I noticed that you're currently using an older version of our Kubernetes provider. Upgrading to version 4 should resolve this issue. The problem you're experiencing stems from the behavior in version 3, where the provider always performs a diff between Pulumi program inputs rather than outputs, which is not the intended behavior you described. This was addressed in [#2445](https://github.com/pulumi/pulumi-kubernetes/pull/2445) by updating the diffing logic to compare the last known cluster output against new inputs, which should resolve the problem. This fix is included in version 4 of our provider.
> 
> Please try upgrading to version 4 of the Pulumi Kubernetes provider and let us know if the issue persists. Thank you!

That's really good news! I will upgrade our Pulumi version within the next week and see if the issue is resolved.

**Comment by loveRhythm1990** [comment]
Pulumi is working properly after we upgrade to `v4.17.1`, thanks again, @rquitales , i will close this issue.
